src/markdown-tests/build.ts

```diff
@@ -0,0 +1,217 @@
+ *  Headlines and codeblocks.
+ */
+function extractTags(markdown: string): (Header | CodeBlock)[] {
+  const result: (Header | CodeBlock)[] = [];
+  const regex = /\n(#+)([^\n]+)\n|```([^`\n]*)\n([^`]*)\n```/gs;
+  let m;
+  while ((m = regex.exec("\n" + markdown)) !== null) {
+    if (m[1] !== undefined) {
+      result.push({
+        kind: "Header",
+        level: m[1].length,
+        title: m[2].trim(),
+      });
+    } else {
+      const title = m[3].trim().split(" ");
+      if (m[3].trim().length === 0) {
+        throw new Error(
+          "Codeblocks are required to have language annotation. If a codeblock should not be included in testing, use `skip` command: `python skip`."
+        );
+      }
+      result.push({
+        kind: "CodeBlock",
+        language: title[0],
+        args: title.slice(1),
+        content: m[4].trim(),
+      });
+    }
+  }
+  return result;
+}
+
+/** Emits a single markdown test. */
+function emitSuite(describe: Describe): string {
+  return `import parse from "frontend/parse";
+import applyLanguage from "@/common/applyLanguage";
+import { findLang } from "@/languages/languages";
+
+function t(name: string, lang: string, input: string, output: string) {
+  test(name, () =>
+    expect(applyLanguage(findLang(lang)!, parse(input))).toEqual(output)
+  );
+}
+
+${emitDescribe(describe)}`;
+}
+
+function emitNode(node: Describe | Test): string {
+  if ("children" in node) return emitDescribe(node);
+  return emitTest(node);
+}
+
+function stringify(x: string): string {
+  return emitStringLiteral(x, [
+    [
+      "`",
+      [
+        [`\\`, `\\\\`],
+        [`\n`, `\\n`],
+        [`\r`, `\\r`],
+        ["`", "\\`"],
+        ["$", "\\$"],
+      ],
+    ],
+    [
+      `"`,
+      [
+        [`\\`, `\\\\`],
+        [`\n`, `\\n`],
+        [`\r`, `\\r`],
+        [`"`, `\\"`],
+      ],
+    ],
+    [
+      `'`,
+      [
+        [`\\`, `\\\\`],
+        [`\n`, `\\n`],
+        [`\r`, `\\r`],
+        [`'`, `\\"`],
+      ],
+    ],
+  ])[0];
+}
+
+function emitDescribe(describe: Describe): string {
+  return (
+    `describe(${stringify(describe.name)}, () => {\n${describe.children
+      .map(emitNode)
+      .join("\n")}`.replaceAll("\n", "\n  ") + "\n});"
+  );
+}
+
+function emitTest(test: Test): string {
+  return `t(${stringify(
+    [test.language, ...test.commands].join(" ")
+  )}, ${stringify(test.language)}, ${stringify(test.input)}, ${stringify(
+    test.output
+  )})`;
+}
+
+/** Compiles a single markdown test suite source to a Jest source. */
+function compileSuite(markdown: string): string {
+  const suite = parseSuite(markdown);
+  if (suite.children.length === 0) return "";
+  return emitSuite(suite);
+}
+
+function readdirSync(p: string, a: string[] = []) {
+  if (fs.statSync(p).isDirectory())
+    fs.readdirSync(p).map((f) =>
+      readdirSync(a[a.push(path.join(p, f)) - 1], a)
+    );
+  return a;
+}
+
+for (const file of readdirSync(process.cwd())) {
```

Comment:
An approach that avoids `node_modules` and works regardless of the path `npm run test:build` is called from. Seems janky though since it depends on the output directory of the node script. Maybe just filter out `node_modules` in `readdirSync`?
```suggestion
for (const file of readdirSync(path.join(__dirname, "../../src"))) {
```
Though idk why I want to filter out node_modules. Some performance instinct ig, but we have no bottleneck here.

Reply:
That came to my mind but I thought depending on the path it is called from makes sense?